app/code/Magento/Catalog/Controller/Adminhtml/Product/Save.php

```diff
@@ -117,6 +117,7 @@ public function execute()
                 $productId = $product->getEntityId();
                 $productAttributeSetId = $product->getAttributeSetId();
                 $productTypeId = $product->getTypeId();
+                $this->getRequest()->setParam('type', $productTypeId);
```

Comment:
Why modyfing request is not a good idea? adding new redirecting option would increase logical complexity which is already to high in this method.